### PR TITLE
Initial NN Architecture, WIP

### PR DESCRIPTION
**Pull request by abhon**
Began working on the NN architecture, have a few things I want to sort out but any feedback appreciated. Not ready for merge.

### CODE REVIEW

**Comment by charlesoblack** [comment]
I'm on Python 3.6. Can we fix to that one instead?

**Comment by abhon** [comment]
Is there anyway we could do both? When it was 3.6 it would just fail since my computer is using 3.7

**Comment by charlesoblack** [comment]
Are you using something specific to 3.7 that we need it?

**Comment by abhon** [comment]
Not really, my computer only has 3.7 on it at the moment. I can redownload 3.6

**Comment by charlesoblack** [comment]
I think you can just install the precommit hook and then modify it back to 3.6 then. I don't think precommit complains.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -3,5 +3,5 @@ repos:
     rev: stable
     hooks:
     - id: black
-      language_version: python3.6
+      language_version: python3.7
```
Comment (charlesoblack) [inline .pre-commit-config.yaml:6]: As mentioned earlier, if you could keep it to 3.6 that would be great lol.

File: alphazero/model.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+class ResidualBlock(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super(ResidualBlock, self).__init__()
+        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=1)
+        self.batch_norm = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(inplace=true)
+
+        def forward(self, x):
+            """
+			Create One Residual Block where we save the input as the residual and add it after two convolutions
```
Comment (charlesoblack) [inline alphazero/model.py:15]: Can you split this into two lines? + dedent line 16

File: alphazero/model.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			"""
+            residual = x
+
+            """
+			first convolution
```
Comment (charlesoblack) [inline alphazero/model.py:20]: triple `"` is meant for docstrings, if you want to comment on stuff you should use `#` instead ;)

File: alphazero/model.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+			"""
+            out = self.conv(x)
+            out = self.batch_norm(out)
+            out = self.relu(out)
+
+            """
+			second convolution, add input after batchnorm step
```
Comment (charlesoblack) [inline alphazero/model.py:27]: same as 20

File: alphazero/model.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+			"""
+            out = self.conv(out)
```
Comment (charlesoblack) [inline alphazero/model.py:29]: I'm not super sure, but I think if you use the same `Conv2d` here you get shared weights across two convolutions. I don't think we want that - I think we want one convolution to have separate weights from the other conv.

File: alphazero/model.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class ResidualBlock(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super(ResidualBlock, self).__init__()
+        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=1)
+        self.batch_norm = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(inplace=true)
```
Comment (charlesoblack) [inline alphazero/model.py:11]: Can you make the batch norm optional? As well as maybe a param for instead of ReLU something like LeakyReLU or something else.

Also, if you're declaring the ReLU to be inplace here, then I think later on (lines 24, 32) you're misusing it, since you're returning the value.

File: alphazero/model.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+			"""
+            out = self.conv(out)
+            out = self.batch_norm(out)
+            out += residual
+            out = self.relu(out)
+
+            return out
+
+
+class Net(nn.Module):
```
Comment (charlesoblack) [inline alphazero/model.py:37]: maybe a better name? :P AlecNet sounds pretty cool imo

File: alphazero/model.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(
+        self,
+        block,
+        in_channels: int,
+        out_channels: int,
+        game_play: Game,
```
Comment (charlesoblack) [inline alphazero/model.py:43]: game_play isn't really a great name for this kwarg, maybe just "game"?

File: alphazero/model.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        residual_layers: int,
```
Comment (charlesoblack) [inline alphazero/model.py:44]: same as 43, maybe num_layers?

File: alphazero/model.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            out_channels=game_play.width * game_play.length + 1,
+        )
+
+        self.valueconv = nn.Conv2D(
+            out_channels, out_channels=1, kernel_size=1, stride=1
+        )
+        self.valuebn = nn.BatchNorm2d(1)
+
+        # figuring out what the in_channel should be for this
+        self.valuefc1 = nn.Linear(in_channels=number, out_channels=256)
```
Comment (charlesoblack) [inline alphazero/model.py:68]: I'm... surprised black allowed this through. number isn't defined here

Comment (charlesoblack) [reply alphazero/model.py:68]: Also, the comment in 67 should maybe be removed?

File: alphazero/model.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class ResidualBlock(nn.Module):
```
Comment (charlesoblack) [inline alphazero/model.py:6]: Can you add docstrings?

File: alphazero/model.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            out += residual
+            out = self.relu(out)
+
+            return out
+
+
+class Net(nn.Module):
+    def __init__(
+        self,
+        block,
```
Comment (charlesoblack) [inline alphazero/model.py:40]: What's block here? I thought it was a counter for how many res blocks to use, but it looks like that's taken care of through num_layers, and block seems to be a callable. (line 75)

File: alphazero/model.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+        layers = []
+        for i in range(0, num_layers):
+            layers.append(block(out_channels, out_channels, stride=1))
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+
+        out = self.conv1(x)
+        out = self.bn(out)
```
Comment (charlesoblack) [inline alphazero/model.py:81]: same as 10/11, could you make batch norm optional? and other nonlinearities?

File: alphazero/model.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    def make_layer(self, block, out_channels, num_layers, stride=1):
+
+        layers = []
+        for i in range(0, num_layers):
+            layers.append(block(out_channels, out_channels, stride=1))
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+
+        out = self.conv1(x)
```
Comment (charlesoblack) [inline alphazero/model.py:80]: I don't really like the name `out` since you use it for both "heads" later. could this maybe be `encoding`?

File: alphazero/model.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        out = self.bn(out)
+        out = self.relu(out)
+
+        out = self.resblock(out)
+
+        policy = self.policyconv(out)
+        policy = self.policybn(policy)
+        policy = self.relu(policy)
+        # a step here I am not completely sure yet
+        policy = self.policyfc1(policy)
```
Comment (charlesoblack) [inline alphazero/model.py:90]: This is also undefined. maybe `policyfc`?

File: alphazero/model.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        self.resblock = self.make_layer(block, out_channels, residual_layers)
+        self.relu = nn.ReLU(inplace=true)
+
+        self.policyconv = nn.Conv2D(
+            out_channels, out_channels=2, kernel_size=2, stride=1
+        )
+        self.policybn = nn.BatchNorm2d(2)
+        self.policyfc = nn.Linear(
+            in_channels=out_channels,
+            out_channels=game_play.width * game_play.length + 1,
```
Comment (charlesoblack) [inline alphazero/model.py:59]: The property you're looking for is `height` not length. Do you add 1 here as a "passing" move? I think that would be game dependent, I think it's possible in Go but not in TTT/C4. Maybe this should be added in `Game` as a property

File: alphazero/model.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+            return out
+
+
+class Net(nn.Module):
+    def __init__(
+        self,
+        block,
+        in_channels: int,
+        out_channels: int,
```
Comment (charlesoblack) [inline alphazero/model.py:42]: this should probably be `encoding_channels` or if you prefer the shorter `enc_channels`, or something like that. Since it's not really the number of output channels

File: alphazero/model.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        game_play: Game,
+        residual_layers: int,
+    ):
+        super(Net, self).__init__()
+
+        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=1)
+        self.bn = nn.BatchNorm2d(out_channels)
+        self.resblock = self.make_layer(block, out_channels, residual_layers)
+        self.relu = nn.ReLU(inplace=true)
```
Comment (charlesoblack) [inline alphazero/model.py:51]: same as above with the `inplace`

File: alphazero/model.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        # a step here I am not completely sure yet
+        policy = self.policyfc1(policy)
+
+        value = self.valueconv(out)
+        value = self.valuebn(value)
+        value = self.relu(value)
+        # a step here I am not completely sure yet
+        value = self.valuefc1(value)
+        value = self.relu(value)
+        value = self.policyfc2(value)
```
Comment (charlesoblack) [inline alphazero/model.py:98]: ? not sure where this is defined either, nor what the policy FCL is doing with the value...

File: alphazero/model.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+        win_value = F.tanh(value)
```
Comment (charlesoblack) [inline alphazero/model.py:99]: Tanh layer instead of using tanh from functional?

File: alphazero/model.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+        return policy, value
```
Comment (charlesoblack) [inline alphazero/model.py:101]: return win_value instead of value?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (abhon) [reply .pre-commit-config.yaml:6]: Fixed.

File: alphazero/model.py (hunk 6 above)
Comment (abhon) [reply alphazero/model.py:11]: I'll work on the params for batch norm/activations later, but I don't really get what you mean by returning the value? The inplace just changes the original tensor without allocating memory for a new tensor. I'll see if it works when I run tests.

File: alphazero/model.py (hunk 7 above)
Comment (abhon) [reply alphazero/model.py:37]: How about SidNet?

File: alphazero/model.py (hunk 6 above)
Comment (charlesoblack) [reply alphazero/model.py:11]: yeah, exactly. later on you act as if you're getting a new tensor as a result from the relu, but you're not.

File: alphazero/model.py (hunk 10 above)
Comment (abhon) [reply alphazero/model.py:68]: Still trying to figure out this step(number is a placeholder). I think there need to be some kind of unpacking here but not completely sure yet. Same with the policy head.

File: alphazero/model.py (hunk 16 above)
Comment (abhon) [reply alphazero/model.py:59]: Yeah the 1 is the passing value. I'll remove it for now, we can find a way to re-implement it if we have a game where passing is a thing.

File: alphazero/model.py (hunk 6 above)
Comment (abhon) [reply alphazero/model.py]: changed inplace to false. Will start running tests later tonight or tomorrow.

**Comment by abhon** [comment]
I've finished the architecture, haven't written a testing script for it yet. However, I'm pretty sure it works(tried things on jupyterlab). I need to just integrate the game class in the tests, which I will do later today.

File: alphazero/model.py [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+        self.logsoftmax = nn.LogSoftmax(dim=1)
+
+    def forward(self, encoding):
+        policy = self.policyconv(encoding)
+        if self.policybn is not None:
+            policy = self.policybn(policy)
+        policy = self.activation(policy)
+        policy = policy.view(-1, self.poss_moves * 2)
+        policy = self.policyfc(policy)
+        probas = self.logsoftmax(policy).exp()
```
Comment (PhilipEkfeldt) [inline alphazero/model.py:134]: It it possible to reshape back before returning here? For easier interaction with game api later

File: alphazero/model.py [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+        batch_on=True,
+        pass_move=True,
+    ):
+
+        super(PolicyHead, self).__init__()
+        self.policyconv = nn.Conv2d(
+            enc_channels, out_channels=2, kernel_size=1, stride=1
+        )
+        self.policybn = nn.BatchNorm2d(2) if batch_on else None
+        self.poss_moves = game.width * game.height
```
Comment (PhilipEkfeldt) [inline alphazero/model.py:113]: This is not necessarily always the case. For TicTacToe yes, for Connect4 no (only 1D vector needed?), for chess it's even more complicated because of selecting which piece to move as well. This is why I don't think we can generalize everything here.

Comment (abhon) [reply alphazero/model.py:113]: I originally named that to be game_area, but in the Go case it was possible moves, sorry about the misleading name. I think we are going have to relate this to the game class: maybe figure out which legal moves are possible and encode them into the 2d game board? Could be a way to do it.

File: alphazero/model.py (hunk 22 above)
Comment (abhon) [reply alphazero/model.py:134]: Will do that!

File: alphazero/model.py [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+    ):
+        super(ResidualBlock, self).__init__()
+        self.conv1 = nn.Conv2d(
+            in_channels, out_channels, kernel_size=3, stride=1, padding=1
+        )
+        self.conv2 = nn.Conv2d(
+            in_channels, out_channels, kernel_size=3, stride=1, padding=1
+        )
+        self.bn1 = nn.BatchNorm2d(out_channels) if batch_on else None
+        self.bn2 = nn.BatchNorm2d(out_channels) if batch_on else None
```
Comment (charlesoblack) [inline alphazero/model.py]: Should probably use nn.Identity() here instead of having the None stuff

File: alphazero/model.py [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+        enc_channels: int,
+        num_blocks: int,
+        activation=nn.ReLU(inplace=False),
+        batch_on=True,
+    ):
+        super(SidNet, self).__init__()
+        self.conv1 = nn.Conv2d(
+            in_channels, enc_channels, kernel_size=3, stride=1, padding=1
+        )
+        self.bn = nn.BatchNorm2d(enc_channels) if batch_on else None
```
Comment (charlesoblack) [inline alphazero/model.py:67]: Same as 19/20

File: alphazero/model.py [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+        activation=nn.ReLU(inplace=False),
+        batch_on=True,
+        pass_move=False,
+    ):
+
+        super(PolicyHead, self).__init__()
+        self.policyconv = nn.Conv2d(
+            enc_channels, out_channels=2, kernel_size=1, stride=1
+        )
+        self.policybn = nn.BatchNorm2d(2) if batch_on else None
```
Comment (charlesoblack) [inline alphazero/model.py:112]: Same as 19/20

File: alphazero/model.py [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+
+        self.game_width = game.width
+        self.game_height = game.height
+        self.board_pos = game.width * game.height
+        self.pass_move = pass_move
+        if self.pass_move:
+            self.policyfc = nn.Linear(
+                in_features=self.board_pos * 2, out_features=self.board_pos + 1,
```
Comment (charlesoblack) [inline alphazero/model.py]: Since `self.pass_move` is supposed to be `True` or `False` you can just add it to `self.board_pos` and it should convert to 1/0 appropriately. no need for this if/else

File: alphazero/model.py [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+        game,
+        activation=nn.ReLU(inplace=False),
+        batch_on=True,
+    ):
+
+        super(ValueHead, self).__init__()
+        self.valueconv = nn.Conv2d(
+            enc_channels, out_channels=1, kernel_size=1, stride=1
+        )
+        self.valuebn = nn.BatchNorm2d(1) if batch_on else None
```
Comment (charlesoblack) [inline alphazero/model.py:164]: Same as 19/20

**Comment by abhon** [comment]
I'll make the combined nn.module tomorrow. Resolved the issues you mentioned yesterday.

**Comment by PhilipEkfeldt** [comment]
I think it looks good now, will have more comments when I'll be testing it with my implementation

**Comment by PhilipEkfeldt** [comment]
Should we merge this? Kinda need this to do my mcts part properly

**Comment by abhon** [comment]
If Sid wants to merge it first, I'm alright with that - I can always start another branch to write the other nn.module. Apologies for not writing it yet, was on the road with family yesterday, trying to catch up with things today.

File: tests/test_model.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+        policy = model.PolicyHead(
+            enc_channels=256, game=self.game, pass_move=False
+        )
+        value = model.ValueHead(enc_channels=256, game=self.game)
+        probas = policy(output)
+        win_value = value(output)
+
+        assert list(probas.size()) == [self.game.width, self.game.height]
+
+        assert list(win_value.size()) == [1, 1]
```
Comment (charlesoblack) [inline tests/test_model.py:43]: Not sure why you convert to list here before comparing. it returns a tensor right?

**Comment by charlesoblack** [comment]
Looks good, only have a question about the comment I just posted. Will you examine this later? I'll go ahead and rebase/merge.

File: tests/test_model.py (hunk 29 above)
Comment (abhon) [reply tests/test_model.py:43]: Yeah I realize now I should have made a tensor rather than a list on the other side of the equality, but it does the same thing.
